### PR TITLE
refactor: removes token check on webui

### PR DESCRIPTION
**Pull request by ayusharma**
**Type:** Refactor

The following has been addressed in the PR:

*  There is a related issue? N/A
*  Unit or Functional tests are included in the PR.

**Description:**

- Removes isTokenExpire flow on webui.
- Removes the `token` and `username` from localStorage on `unauthorised` 
- Logouts the user as web ui receives 401 from backend API. 
- Removes Unused base-64 utilites.

### CODE REVIEW

File: src/webui/utils/api.js [hunk 1]
```diff
@@ -2,47 +2,54 @@ import storage from './storage';
...
+    if (!window.VERDACCIO_API_URL) {
+      throw new Error('VERDACCIO_API_URL is not defined!');
+    }
 
-      const token = storage.getItem('token');
-      if (token) {
-        if (!options.headers) options.headers = {};
+    const token = storage.getItem('token');
+    if (token) {
+      if (!options.headers) options.headers = {};
```
Comment (juanpicado) [inline src/webui/utils/api.js]:
```
if (!options.headers) {
   options.headers = {};
}
```

File: src/webui/utils/api.js [hunk 2]
```diff
@@ -2,47 +2,54 @@ import storage from './storage';
...
-          credentials: 'same-origin',
-          ...options
+    return new Promise((resolve, reject) => {
+      fetch(url, {
+        method,
+        credentials: 'same-origin',
+        ...options
+      })
+        .then(function checkAuth(response) {
+          if (response.status === 401) {
```
Comment (juanpicado) [inline src/webui/utils/api.js]: Use constants for `401`

File: src/webui/utils/api.js [hunk 3]
```diff
@@ -1,59 +1,104 @@
 import storage from './storage';
```
Comment (ayusharma) [inline src/webui/utils/api.js]: @juanpicado I have done changes here. API service refactor. Please review.

File: src/webui/utils/api.js [hunk 4]
```diff
@@ -1,59 +1,104 @@
...
-        if (!options.headers) options.headers = {};
+/**
+ * Handles response according to content type
+ * @param {object} response
+ * @returns {promise}
+ */
+function handleResponseType(response) {
+  if (response.headers) {
+    const contentType = response.headers.get('Content-Type');
+    if (contentType.includes('application/pdf')) {
```
Comment (juanpicado) [inline src/webui/utils/api.js]: `'application/pdf'`? do we need this?

File: src/webui/utils/api.js [hunk 5]
```diff
@@ -1,59 +1,104 @@
...
-        options.headers.authorization = `Bearer ${token}`;
-      }
+/**
+ * checkAuthentication if 401 then clears storage
+ */
+function checkAuthentication(response) {
+  if (response.status === HTTP_STATUS.UNAUTHORIZED) {
+    storage.removeItem('token');
+    storage.removeItem('username');
+    alert('You\'ve been logged out from server');
```
Comment (juanpicado) [inline src/webui/utils/api.js]: I'm not sure whether we have to notify the user about this.

File: src/webui/utils/api.js (hunk 4 above)
Comment (ayusharma) [reply src/webui/utils/api.js]: No.. 😉  I'll remove it

File: src/webui/utils/api.js [hunk 6]
```diff
@@ -1,49 +1,84 @@
...
-      }
+/**
+ * checkAuthentication if 401 then clears storage
+ */
+function checkAuthentication(response) {
+  if (response.status === HTTP_STATUS.UNAUTHORIZED) {
+    storage.removeItem('token');
+    storage.removeItem('username');
+    /**
+     * @todo
```
Comment (ayusharma) [inline src/webui/utils/api.js]: @juanpicado As discussed.

File: src/webui/app.js [hunk 7]
```diff
@@ -52,8 +52,9 @@ export default class App extends Component {
     // checks for token validity
```
Comment (juanpicado) [inline src/webui/app.js]: This comment is not needed I guess.

File: src/webui/utils/api.js [hunk 8]
```diff
@@ -1,49 +1,84 @@
...
-      if (!window.VERDACCIO_API_URL) {
-        throw new Error('VERDACCIO_API_URL is not defined!');
-      }
+/**
+ * CHECK FOR BASE API URL
+ */
+if (!window.VERDACCIO_API_URL) {
+  throw new Error('VERDACCIO_API_URL is not defined!');
+}
+const BASE_URL = window.VERDACCIO_API_URL;
```
Comment (juanpicado) [inline src/webui/utils/api.js]: I'd move this above the previous condition and use `BASE_URL` with the `if` condition. So, easy to change in the future.

File: src/webui/utils/api.js [hunk 9]
```diff
@@ -1,49 +1,84 @@
...
-      const token = storage.getItem('token');
-      if (token) {
-        if (!options.headers) options.headers = {};
+/**
+ * Handles response according to content type
+ * @param {object} response
+ * @returns {promise}
+ */
+function handleResponseType(response) {
+  if (response.headers) {
```
Comment (juanpicado) [inline src/webui/utils/api.js]: You might destructure `response` here as follows:

```
const {headers, ok, json, text} = response;
``

File: src/webui/utils/api.js (hunk 6 above)
Comment (juanpicado) [reply src/webui/utils/api.js]: Agree, for now, we will reload the page, once we move to a better state manager we will do something different.

File: src/webui/utils/api.js [hunk 10]
```diff
@@ -1,49 +1,84 @@
...
+    const contentType = response.headers.get('Content-Type');
+    if (contentType.includes('application/json')) {
```
Comment (juanpicado) [inline src/webui/utils/api.js]: `'application/json'` I think we have constants for this.

File: src/webui/utils/api.js [hunk 11]
```diff
@@ -1,49 +1,84 @@
...
+};
 
-        options.headers.authorization = `Bearer ${token}`;
-      }
+/**
+ * checkAuthentication if 401 then clears storage
+ */
+function checkAuthentication(response) {
+  if (response.status === HTTP_STATUS.UNAUTHORIZED) {
+    storage.removeItem('token');
```
Comment (juanpicado) [inline src/webui/utils/api.js]: `'token'` and `'username' might be constants at the top, for future maintenance. 

```
const LOCAL_STORAGE_TOKEN = 'token';
const LOCAL_STORAGE_USERNAME = 'username'
```

File: src/webui/utils/api.js [hunk 12]
```diff
@@ -1,49 +1,84 @@
...
+  request(resource = '', method = 'GET', options = {}) {
+    let endpoint;
+    
+    // checks for token
+    const token = storage.getItem('token');
+    if (token) {
+      if (!options.headers) {
+        options.headers = {};
       }
+      options.headers.authorization = `Bearer ${token}`;
```
Comment (juanpicado) [inline src/webui/utils/api.js]: `Bearer` already exist as constant `TOKEN_BEARER`

We have a utility to build this ``Bearer ${token}`` find for
```
export function buildToken(type: string, token: string) {
  return `${_.capitalize(type)} ${token}`;
}
```

File: src/webui/utils/api.js [hunk 13]
```diff
@@ -1,49 +1,84 @@
...
-        })
+    // Checks if resource is already a URL
+    const protocols = ['http://', 'https://', '//'];
+    const isResource = protocols.some((prefix) => resource.startsWith(prefix));
+    if (isResource === false) {
+      endpoint = this.baseUrl + resource;
+    }
+
+    // Final Promise
+    return new Promise(function APIPromise(resolve, reject){
```
Comment (juanpicado) [inline src/webui/utils/api.js]: `eslint` should enforce `){` space between those characters.

File: test/unit/setup.js [hunk 14]
```diff
@@ -18,3 +18,4 @@ if (global.document) {
   }));
   document.execCommand = jest.fn();
 }
+global.VERDACCIO_API_URL = 'http://localhost/-/verdaccio/'
```
Comment (juanpicado) [inline test/unit/setup.js]: Here `global.VERDACCIO_API_URL` I'd use something different as `global.TEST_VERDACCIO_API_URL` then when you search by `VERDACCIO_API_URL` we don't confuse with `window.VERDACCIO_API_URL`

File: test/unit/webui/app.spec.js [hunk 15]
```diff
@@ -54,16 +54,21 @@ describe('App', () => {
   });
 
   it('isUserAlreadyLoggedIn: token already available in storage', async () => {
-
     storage.setItem('username', 'verdaccio');
-    storage.setItem('token', generateTokenWithTimeRange(24));
+    storage.setItem('token', 'TEST_TOKEN');
     const { isUserAlreadyLoggedIn } = wrapper.instance();
-
+    const result = {token: 'TEST_TOKEN', username: 'verdaccio'};
```
Comment (juanpicado) [inline test/unit/webui/app.spec.js]: `'TEST_TOKEN' is twice, so, `const`

File: test/unit/webui/app.spec.js [hunk 16]
```diff
@@ -54,16 +54,21 @@ describe('App', () => {
...
     isUserAlreadyLoggedIn();
+    expect(wrapper.state('user')).toEqual(result);
+  });
 
-    expect(wrapper.state('user').username).toEqual('verdaccio');
+  it('isUserAlreadyLoggedIn: token not available in storage', async () => {
```
Comment (juanpicado) [inline test/unit/webui/app.spec.js]: `async` is not necessary here I guess.

File: test/unit/webui/app.spec.js [hunk 17]
```diff
@@ -36,6 +34,8 @@ describe('App', () => {
 
   beforeEach(() => {
     wrapper = mount(<App />);
+    storage.removeItem('username');
+    storage.removeItem('token');
```
Comment (juanpicado) [inline test/unit/webui/app.spec.js]: I'd rather use `storage.clear()`, easier for future new keys.

File: test/unit/webui/app.spec.js [hunk 18]
```diff
@@ -36,6 +34,8 @@ describe('App', () => {
 
   beforeEach(() => {
     wrapper = mount(<App />);
```
Comment (juanpicado) [inline test/unit/webui/app.spec.js]: Do we need `mount`?

File: test/unit/webui/app.spec.js [hunk 19]
```diff
@@ -54,16 +54,21 @@ describe('App', () => {
   });
 
   it('isUserAlreadyLoggedIn: token already available in storage', async () => {
-
     storage.setItem('username', 'verdaccio');
```
Comment (juanpicado) [inline test/unit/webui/app.spec.js]: `'token'` and `'username` if are constants, we can reuse them here.

File: test/unit/webui/app.spec.js (hunk 15 above)
Comment (juanpicado) [reply test/unit/webui/app.spec.js]: `const result = {token: 'TEST_TOKEN', username: 'verdaccio'};` I would suggest the local storage should return this object, is more reliable the shape match properly.

File: src/webui/utils/api.js (hunk 13 above)
Comment (juanpicado) [reply src/webui/utils/api.js]: or prettify, in any case, we have to check that in the future eslint PR. So, just check but don't change it. I want this fails when I apply the proper rules.

**Pull request closed without merging**